In [ ]:
# Murphy Campbell
# email: murphybcampbell@googlemail.com

# TTA Data Academy Home Learning: Session: !2 Jan 2022

# Reference: 
# https://en.wikibooks.org/wiki/SQL_Exercises/Employee_management#Sample_dataset
# https://jupyter-tutorial.readthedocs.io/en/stable/data-processing/postgresql/ipython-sql.html
# https://github.com/catherinedevlin/ipython-sql

In [1]:
# Install mysql server
!apt-get install mysql-server > /dev/null

In [2]:
!service mysql start

 * Starting MySQL database server mysqld
No directory, logging in with HOME=/
   ...done.


In [ ]:
# !sudo mysql_secure_installation

In [3]:
# Setup user mysqluser and grant full privileges 
!mysql

Welcome to the MySQL monitor.  Commands end with ; or \g.
Your MySQL connection id is 4
Server version: 5.7.36-0ubuntu0.18.04.1 (Ubuntu)

Copyright (c) 2000, 2021, Oracle and/or its affiliates.

Oracle is a registered trademark of Oracle Corporation and/or its
affiliates. Other names may be trademarks of their respective
owners.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

mysql> quit
Bye


In [6]:
# Install mysql python client driver module
!pip install mysql

     |████████████████████████████████| 87 kB 3.9 MB/s 
  Created wheel for mysqlclient: filename=mysqlclient-2.1.0-cp37-cp37m-linux_x86_64.whl size=99977 sha256=29328ed9dc2eaf52c26cfe0ae09c8f8676d418d311ff99108628efc5725e0b1d
  Stored in directory: /root/.cache/pip/wheels/97/d4/df/08cd6e1fa4a8691b268ab254bd0fa589827ab5b65638c010b4
Successfully built mysqlclient


In [4]:
# Create user mysqluser with password using the mysql client
!mysql

Welcome to the MySQL monitor.  Commands end with ; or \g.
Your MySQL connection id is 5
Server version: 5.7.36-0ubuntu0.18.04.1 (Ubuntu)

Copyright (c) 2000, 2021, Oracle and/or its affiliates.

Oracle is a registered trademark of Oracle Corporation and/or its
affiliates. Other names may be trademarks of their respective
owners.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

mysql> CREATE USER ' mysqluser'@'localhost' IDENTIFIED BY '43JmXApHdF7';
Query OK, 0 rows affected (0.00 sec)

mysql> GRANT ALL PRIVILEGES ON *.* TO 'mysqluser'@'localhost' WITH GRANT OPTION; 
Query OK, 0 rows affected (0.00 sec)

mysql> quit
Bye


In [7]:
import sqlalchemy


In [10]:
# Create connection to mysql database instance
sqlalchemy.create_engine('mysql://mysqluser:43JmXApHdF7@127.0.0.1')

Engine(mysql://mysqluser:***@127.0.0.1)

In [11]:
# Load sql extension
%load_ext sql


The sql extension is already loaded. To reload it, use:
  %reload_ext sql
(MySQLdb._exceptions.OperationalError) (1049, "Unknown database 'mydb'")
(Background on this error at: https://sqlalche.me/e/14/e3q8)
Connection info needed in SQLAlchemy format, example:
               postgresql://username:password@hostname/dbname
               or an existing connection: dict_keys([])


In [12]:
# Connect to mysql database instance using user mysqluser with password
%sql mysql://mysqluser:43JmXApHdF7@localhost

'Connected: mysqluser@None'

In [13]:
# Create database mydb
%sql create database mydb 

 * mysql://mysqluser:***@localhost
1 rows affected.


[]

In [14]:
# Attach to mydb instance
%sql use mydb

 * mysql://mysqluser:***@localhost
0 rows affected.


[]

In [16]:
%%sql 
show databases

 * mysql://mysqluser:***@localhost
5 rows affected.


[('information_schema',),
 ('mydb',),
 ('mysql',),
 ('performance_schema',),
 ('sys',)]

In [18]:
# Create table Departments
%%sql
CREATE TABLE Departments (
Code INT(4) PRIMARY KEY,
Name varchar(64) NOT NULL ,
Budget DECIMAL(16,2) NOT NULL
);


 * mysql://mysqluser:***@localhost
0 rows affected.


[]

In [19]:
# Create table Employee
%%sql
CREATE TABLE Employees (
SSN INT (16) PRIMARY KEY,
FirstName varchar(24) NOT NULL ,
LastName varchar(32) NOT NULL ,
Department INT (4) NOT NULL ,
MobileNumber INT (16) NOT NULL ,
foreign key (department) references Departments(Code)
) ENGINE=INNODB;

 * mysql://mysqluser:***@localhost
0 rows affected.


[]

In [20]:
%sql show tables

 * mysql://mysqluser:***@localhost
2 rows affected.


[('Departments',), ('Employees',)]

View and show both table structures and data to make sure they are setup correctly.

In [28]:
%sql describe Employees;

 * mysql://mysqluser:***@localhost
5 rows affected.


[('SSN', 'int(16)', 'NO', 'PRI', None, ''),
 ('FirstName', 'varchar(24)', 'NO', '', None, ''),
 ('LastName', 'varchar(32)', 'NO', '', None, ''),
 ('Department', 'int(4)', 'NO', 'MUL', None, ''),
 ('MobileNumber', 'int(32)', 'YES', '', None, '')]

In [29]:
%sql describe Departments;


 * mysql://mysqluser:***@localhost
3 rows affected.


[('Code', 'int(4)', 'NO', 'PRI', None, ''),
 ('Name', 'varchar(64)', 'NO', '', None, ''),
 ('Budget', 'decimal(16,2)', 'NO', '', None, '')]

In [96]:
%%sql
ALTER TABLE Employees
CHANGE COLUMN MobileNumber MobileNumber INT(16) NOT NULL;


 * mysql://mysqluser:***@localhost
0 rows affected.


[]

In [97]:
%%sql
describe Employees

 * mysql://mysqluser:***@localhost
5 rows affected.


[('SSN', 'int(16)', 'NO', 'PRI', None, ''),
 ('FirstName', 'varchar(24)', 'NO', '', None, ''),
 ('LastName', 'varchar(32)', 'NO', '', None, ''),
 ('Department', 'int(4)', 'NO', 'MUL', None, ''),
 ('MobileNumber', 'int(16)', 'NO', '', None, '')]

In [30]:
%%sql
delete from Departments

 * mysql://mysqluser:***@localhost
10 rows affected.


[]

In [31]:
%%sql
select * from Departments

 * mysql://mysqluser:***@localhost
0 rows affected.


[]

Enter records into both tables and view them. Make sure at least one of your tables has
10 records.

In [32]:


%%sql
INSERT INTO Departments VALUES
(14,'IT',635000), (37,'Accounting',15000),
(51,' Purchasing',235000), (59,'Human Resources',240000),
(77,'Research',55000), (79,'Technical',55000),
(43,'Customer Service',34500), (83,'Communications',32000),
(95,'Internation Affairs',924500), (63,'Exports',832000);


 * mysql://mysqluser:***@localhost
10 rows affected.


[]

Load data into Employees table

In [125]:
%%sql
INSERT INTO Employees VALUES
('123234877','Michael','Rogers',14,0976543909),
('152934485','Anand','Manikutty',14,0787643989),
('546523478','John','Doe',59,675394394),
('326587417','Joe','Stevens',37,0566432900),
('332569843','George','O Donnell',77,874),
('631231482','David','Smith',77,4338),
('654873219','Zacary','Efron',59,643809),
('745685214','Eric','Goldsmith',59,0895447),
('845657245','Elizabeth','Doe',14,7553086),
('332154719','Mary-Anne','Foster',14,34228333), 
('845657246','Kumar','Swamy',14,08326522),
('222364883','Carol','Smith',37,0134533);








 * mysql://mysqluser:***@localhost
12 rows affected.


[]

In [126]:
%%sql
select * from Departments;

 * mysql://mysqluser:***@localhost
10 rows affected.


[(14, 'IT', Decimal('635000.00')),
 (37, 'Accounting', Decimal('15000.00')),
 (43, 'Customer Service', Decimal('34500.00')),
 (51, ' Purchasing', Decimal('235000.00')),
 (59, 'Human Resources', Decimal('240000.00')),
 (63, 'Exports', Decimal('832000.00')),
 (77, 'Research', Decimal('55000.00')),
 (79, 'Technical', Decimal('55000.00')),
 (83, 'Communications', Decimal('32000.00')),
 (95, 'Internation Affairs', Decimal('924500.00'))]

In [127]:
%%sql
select * from Employees;

 * mysql://mysqluser:***@localhost
12 rows affected.


[(123234877, 'Michael', 'Rogers', 14, 976543909),
 (152934485, 'Anand', 'Manikutty', 14, 787643989),
 (222364883, 'Carol', 'Smith', 37, 134533),
 (326587417, 'Joe', 'Stevens', 37, 566432900),
 (332154719, 'Mary-Anne', 'Foster', 14, 34228333),
 (332569843, 'George', 'O Donnell', 77, 874),
 (546523478, 'John', 'Doe', 59, 675394394),
 (631231482, 'David', 'Smith', 77, 4338),
 (654873219, 'Zacary', 'Efron', 59, 643809),
 (745685214, 'Eric', 'Goldsmith', 59, 895447),
 (845657245, 'Elizabeth', 'Doe', 14, 7553086),
 (845657246, 'Kumar', 'Swamy', 14, 8326522)]

Delete record John Doe from the Employees table

In [128]:
%%sql
delete from Employees where FirstName='John' AND LastName='Doe' ;

 * mysql://mysqluser:***@localhost
1 rows affected.


[]

Check that the record John Doe has been deleted

In [130]:
%%sql
select * from Employees where FirstName='John' AND LastName='Doe' ;

 * mysql://mysqluser:***@localhost
0 rows affected.


[]


Selecting record Carol Smith prior to updating her
Mobile Number from 134533 to 999999999

In [132]:
%%sql
select * from Employees where FirstName='Carol' AND LastName='Smith' ;

 * mysql://mysqluser:***@localhost
1 rows affected.


[(222364883, 'Carol', 'Smith', 37, 134533)]

Updating her Mobile Number

In [133]:
%%sql
UPDATE Employees
SET MobileNumber = '999999999'
WHERE FirstName='Carol' AND LastName='Smith' ;

 * mysql://mysqluser:***@localhost
1 rows affected.


[]

Check the update was carried out

In [134]:
%%sql
select * from Employees where FirstName='Carol' AND LastName='Smith' ;

 * mysql://mysqluser:***@localhost
1 rows affected.


[(222364883, 'Carol', 'Smith', 37, 999999999)]

Join tables using cross join

In [140]:
%%sql
select FirstName, Name from Employees CROSS JOIN Departments where Name='IT' ORDER by FirstName;

 * mysql://mysqluser:***@localhost
11 rows affected.


[('Anand', 'IT'),
 ('Carol', 'IT'),
 ('David', 'IT'),
 ('Elizabeth', 'IT'),
 ('Eric', 'IT'),
 ('George', 'IT'),
 ('Joe', 'IT'),
 ('Kumar', 'IT'),
 ('Mary-Anne', 'IT'),
 ('Michael', 'IT'),
 ('Zacary', 'IT')]

In [157]:
%%sql
select Employees.Department, Departments.Code from Employees inner join Departments
on Employees.Department=Departments.Code;

 * mysql://mysqluser:***@localhost
11 rows affected.


[(14, 14),
 (14, 14),
 (14, 14),
 (14, 14),
 (14, 14),
 (37, 37),
 (37, 37),
 (59, 59),
 (59, 59),
 (77, 77),
 (77, 77)]

In [185]:
%%sql
select MobileNumber, Code from Employees full join Departments where MobileNumber > 787643989; 
 

 * mysql://mysqluser:***@localhost
20 rows affected.


[(976543909, 14),
 (999999999, 14),
 (976543909, 37),
 (999999999, 37),
 (976543909, 43),
 (999999999, 43),
 (976543909, 51),
 (999999999, 51),
 (976543909, 59),
 (999999999, 59),
 (976543909, 63),
 (999999999, 63),
 (976543909, 77),
 (999999999, 77),
 (976543909, 79),
 (999999999, 79),
 (976543909, 83),
 (999999999, 83),
 (976543909, 95),
 (999999999, 95)]

In [189]:
%%sql
select MobileNumber, Code from Employees inner join Departments where MobileNumber = 787643989; 

 * mysql://mysqluser:***@localhost
10 rows affected.


[(787643989, 14),
 (787643989, 37),
 (787643989, 43),
 (787643989, 51),
 (787643989, 59),
 (787643989, 63),
 (787643989, 77),
 (787643989, 79),
 (787643989, 83),
 (787643989, 95)]

A Simple query

In [193]:
%%sql
select FirstName from Employees order by Firstname; 

 * mysql://mysqluser:***@localhost
11 rows affected.


[('Anand',),
 ('Carol',),
 ('David',),
 ('Elizabeth',),
 ('Eric',),
 ('George',),
 ('Joe',),
 ('Kumar',),
 ('Mary-Anne',),
 ('Michael',),
 ('Zacary',)]

Some complex query (more than one field/column) to demonstrate the relations
between the 2 tables

In [206]:
%%sql
select Employees.* from Employees where Department IN (select Code from Departments where Code in 
(select Code from Departments where Budget > 190000) );


 * mysql://mysqluser:***@localhost
7 rows affected.


[(123234877, 'Michael', 'Rogers', 14, 976543909),
 (152934485, 'Anand', 'Manikutty', 14, 787643989),
 (332154719, 'Mary-Anne', 'Foster', 14, 34228333),
 (654873219, 'Zacary', 'Efron', 59, 643809),
 (745685214, 'Eric', 'Goldsmith', 59, 895447),
 (845657245, 'Elizabeth', 'Doe', 14, 7553086),
 (845657246, 'Kumar', 'Swamy', 14, 8326522)]

In [216]:
%%sql
select Departments.* from Departments where Budget IN (select Budget from Departments where Budget > 190000)
order by Budget;


 * mysql://mysqluser:***@localhost
5 rows affected.


[(51, ' Purchasing', Decimal('235000.00')),
 (59, 'Human Resources', Decimal('240000.00')),
 (14, 'IT', Decimal('635000.00')),
 (63, 'Exports', Decimal('832000.00')),
 (95, 'Internation Affairs', Decimal('924500.00'))]

Retrieve all your data sorted in ascending order on an appropriate field (one table)



In [217]:
%%sql
select * from Employees order by LastName;

 * mysql://mysqluser:***@localhost
11 rows affected.


[(845657245, 'Elizabeth', 'Doe', 14, 7553086),
 (654873219, 'Zacary', 'Efron', 59, 643809),
 (332154719, 'Mary-Anne', 'Foster', 14, 34228333),
 (745685214, 'Eric', 'Goldsmith', 59, 895447),
 (152934485, 'Anand', 'Manikutty', 14, 787643989),
 (332569843, 'George', 'O Donnell', 77, 874),
 (123234877, 'Michael', 'Rogers', 14, 976543909),
 (222364883, 'Carol', 'Smith', 37, 999999999),
 (631231482, 'David', 'Smith', 77, 4338),
 (326587417, 'Joe', 'Stevens', 37, 566432900),
 (845657246, 'Kumar', 'Swamy', 14, 8326522)]

Filter data using comparison operators (one table)

In [227]:
%%sql
select * from Employees where (Department = 14 ) and (FirstName = 'Anand');

 * mysql://mysqluser:***@localhost
1 rows affected.


[(152934485, 'Anand', 'Manikutty', 14, 787643989)]

In [235]:
%%sql
select * from Employees where ((Department >= 14 ) and (MobileNumber > 8473383)) order by MobileNumber;


 * mysql://mysqluser:***@localhost
5 rows affected.


[(332154719, 'Mary-Anne', 'Foster', 14, 34228333),
 (326587417, 'Joe', 'Stevens', 37, 566432900),
 (152934485, 'Anand', 'Manikutty', 14, 787643989),
 (123234877, 'Michael', 'Rogers', 14, 976543909),
 (222364883, 'Carol', 'Smith', 37, 999999999)]